mia/attack/estimators.py

```diff
@@ -1,67 +1,111 @@
+    - attack (BaseEstimator): Trained attack model.
+
+    Methods:
+    - fit(X: ArrayLike, y: ArrayLike) -> BaseEstimator: Fit the attack model using the specified dataset.
+    - predict(X: ArrayLike) -> ArrayLike: Make predictions using the trained attack model.
+    - score(X: ArrayLike, y: ArrayLike) -> ArrayLike: Score the attack model on the provided dataset.
+
+    Constants:
+    - SPLIT_SIZE (float): Default split size for shadow models, set to 0.25.
+    """
+
+    SPLIT_SIZE = 0.25
+
+    def __init__(
+        self,
+        factory: typing.Callable[[], BaseEstimator],
+        categories: int,
+        shadows: int = 20,
+    ):
+        """
+        Initialize the MIA instance.
+
+        Parameters:
+        - factory (typing.Callable[[], BaseEstimator]): A callable that creates an instance of the target model.
+        - categories (int): Number of categories in the target model's output.
+        - shadows (int, optional): Number of shadow models to train. Defaults to 20.
+        """
         self.factory = factory
         self.shadows = shadows
+        self.categories = categories
+
+    def fit(self, X: ArrayLike, y: ArrayLike) -> BaseEstimator:
+        """
+        Fit the MIA attack model using the specified dataset.
+
+        Parameters:
+        - X (ArrayLike): Input features.
+        - y (ArrayLike): Target labels.
+
+        Returns:
+        - self (BaseEstimator): Fitted MIA instance.
+        """
+        iterator = ShuffleSplit(
+            n_splits=self.shadows,
+            test_size=self.SPLIT_SIZE,
+            train_size=self.SPLIT_SIZE,
+        )
 
-    def fit(self, X: ArrayLike, y: ArrayLike) -> Shadow:
-        self.models = []
-        self.inside = []
-        self.outside = []
+        X_shadow = numpy.empty((0, self.categories))
+        y_shadow = numpy.empty((0,))
 
-        iterator = ShuffleSplit(n_splits=self.shadows, test_size=0.5)
+        for inside_index, outside_index in iterator.split(X):
+            X_train, y_train = X[inside_index], y[inside_index]
+            model = self.factory().fit(X_train, y_train)
 
-        for inside, outside in iterator.split(X):
-            X_train, y_train = X[inside], y[inside]
+            inside_predictions = model.predict_proba(X[inside_index])
+            outside_predictions = model.predict_proba(X[outside_index])
 
-            model = self.factory()
-            model.fit(X_train, y_train)
+            inside_labels = numpy.zeros(inside_predictions.shape[0])
+            outside_labels = numpy.ones(outside_predictions.shape[0])
 
-            self.models.append(model)
-            self.inside.append(inside)
-            self.outside.append(outside)
+            X_shadow = numpy.vstack([X_shadow, inside_predictions, outside_predictions])
+            y_shadow = numpy.hstack([y_shadow, inside_labels, outside_labels])
 
-        self.X_fit_ = X
-        self.y_fit_ = y
+        self.attack = self.factory().fit(X_shadow, y_shadow)
```

Comment:
does it mean that the attack model is only trained with (prediction, in/out) instead of (prediction, label, in/out)? So that we have one single attack model that trained on dataset generated by all k shadow models?

Reply:
Yes, the model trained like that performs almost as good as expected by the paper results. So I decided to go that route instead of making the bundle model.